File: frontend/src/engine/ipIntelligence.js
/**
 * IP Intelligence Module
 * Performs IP geolocation, VPN/proxy detection, and cloud provider identification.
 */

/**
 * Look up IP geolocation via ip-api.com
 */
export async function lookupIP(ip) {
  try {
    const response = await fetch(`http://ip-api.com/json/${ip}?fields=status,message,country,countryCode,region,regionName,city,zip,lat,lon,timezone,isp,org,as,asname,reverse,mobile,proxy,hosting,query`);
    const data = await response.json();

    if (data.status === 'success') {
      return {
        ip: data.query,
        country: data.country,
        countryCode: data.countryCode,
        region: data.regionName,
        city: data.city,
        lat: data.lat,
        lon: data.lon,
        timezone: data.timezone,
        isp: data.isp,
        org: data.org,
        as: data.as,
        asName: data.asname,
        reverse: data.reverse,
        isMobile: data.mobile,
        isProxy: data.proxy,
        isHosting: data.hosting,
        riskIndicators: assessIPRisk(data),
        error: null,
      };
    } else {
      return createFallbackResult(ip, data.message);
    }
  } catch (error) {
    return createFallbackResult(ip, error.message);
  }
}

/**
 * Batch IP lookup with rate limiting
 */
export async function batchLookupIPs(ips) {
  const results = [];

  for (let i = 0; i < ips.length; i++) {
    // Rate limit: ip-api allows 45 requests per minute
    if (i > 0) {
      await new Promise(resolve => setTimeout(resolve, 1500));
    }
    const result = await lookupIP(ips[i]);
    results.push(result);
  }

  return results;
}

/**
 * Assess IP risk based on geolocation data
 */
function assessIPRisk(data) {
  const indicators = [];

  // Proxy/VPN detection
  if (data.proxy) {
    indicators.push({
      type: 'PROXY_VPN',
      severity: 'high',
      description: 'IP is associated with a proxy, VPN, or anonymization service',
    });
  }

  // Hosting/datacenter IP
  if (data.hosting) {
    indicators.push({
      type: 'HOSTING_PROVIDER',
      severity: 'medium',
      description: `IP belongs to a hosting/cloud provider (${data.org || data.isp})`,
    });
  }

  // Known cloud providers
  const cloudProviders = detectCloudProvider(data.org || '', data.isp || '', data.as || '');
  if (cloudProviders) {
    indicators.push({
      type: 'CLOUD_INFRASTRUCTURE',
      severity: 'medium',
      description: `IP hosted on ${cloudProviders} cloud infrastructure`,
    });
  }

  // TOR exit node detection (heuristic)
  if (isTORIndicator(data.org || '', data.isp || '', data.reverse || '')) {
    indicators.push({
      type: 'TOR_EXIT',
      severity: 'critical',
      description: 'IP appears to be a TOR exit node — anonymized traffic',
    });
  }

  // High-risk country (configurable)
  const highRiskCountries = ['RU', 'CN', 'KP', 'IR', 'NG', 'RO', 'UA', 'BY'];
  if (highRiskCountries.includes(data.countryCode)) {
    indicators.push({
      type: 'HIGH_RISK_COUNTRY',
      severity: 'medium',
      description: `IP originates from high-risk region: ${data.country}`,
    });
  }

  return indicators;
}

/**
 * Detect cloud provider from ASN/org data
 */
function detectCloudProvider(org, isp, as) {
  const lowerOrg = (org + ' ' + isp + ' ' + as).toLowerCase();

  if (lowerOrg.includes('amazon') || lowerOrg.includes('aws') || lowerOrg.includes('ec2')) return 'Amazon Web Services (AWS)';
  if (lowerOrg.includes('microsoft') || lowerOrg.includes('azure')) return 'Microsoft Azure';
  if (lowerOrg.includes('google') || lowerOrg.includes('gcp') || lowerOrg.includes('cloud')) return 'Google Cloud Platform';
  if (lowerOrg.includes('digitalocean')) return 'DigitalOcean';
  if (lowerOrg.includes('linode') || lowerOrg.includes('akamai')) return 'Linode/Akamai';
  if (lowerOrg.includes('ovh')) return 'OVH';
  if (lowerOrg.includes('hetzner')) return 'Hetzner';
  if (lowerOrg.includes('vultr')) return 'Vultr';
  if (lowerOrg.includes('cloudflare')) return 'Cloudflare';

  return null;
}

/**
 * Check for TOR-related indicators
 */
function isTORIndicator(org, isp, reverse) {
  const combined = (org + ' ' + isp + ' ' + reverse).toLowerCase();
  return combined.includes('tor') || combined.includes('exit') || combined.includes('relay');
}

/**
 * Create fallback result for failed lookups
 */
function createFallbackResult(ip, errorMessage) {
  return {
    ip,
    country: 'Unknown',
    countryCode: 'XX',
    region: 'Unknown',
    city: 'Unknown',
    lat: 0,
    lon: 0,
    timezone: '',
    isp: 'Unknown',
    org: 'Unknown',
    as: '',
    asName: '',
    reverse: '',
    isMobile: false,
    isProxy: false,
    isHosting: false,
    riskIndicators: [],
    error: errorMessage,
  };
}

/**
 * Calculate IP risk score
 */
export function calculateIPRiskScore(ipResult) {
  if (!ipResult || ipResult.error) return 30;

  let score = 0;

  for (const indicator of ipResult.riskIndicators) {
    switch (indicator.severity) {
      case 'critical': score += 40; break;
      case 'high': score += 30; break;
      case 'medium': score += 15; break;
      case 'low': score += 5; break;
    }
  }

  return Math.min(100, score);
}

/**
 * Get country flag emoji from country code
 */
export function getCountryFlag(countryCode) {
  if (!countryCode || countryCode === 'XX') return '🌍';
  const codePoints = countryCode
    .toUpperCase()
    .split('')
    .map(char => 127397 + char.charCodeAt(0));
  return String.fromCodePoint(...codePoints);
}

File: backend/engine/ip_intelligence.py
"""
IP Intelligence Module
Performs IP geolocation lookups via ip-api.com, proxy/hosting detection,
cloud infrastructure classification, and risk evaluation.
"""

import httpx
import asyncio

async def lookup_ip(ip: str) -> dict:
    url = f"http://ip-api.com/json/{ip}?fields=status,message,country,countryCode,region,regionName,city,zip,lat,lon,timezone,isp,org,as,asname,reverse,mobile,proxy,hosting,query"
    try:
        async with httpx.AsyncClient(timeout=4.0) as client:
            response = await client.get(url)
            if response.status_code == 200:
                data = response.json()
                if data.get("status") == "success":
                    return {
                        "ip": data.get("query", ip),
                        "country": data.get("country", "Unknown"),
                        "countryCode": data.get("countryCode", "XX"),
                        "region": data.get("regionName", "Unknown"),
                        "city": data.get("city", "Unknown"),
                        "lat": data.get("lat", 0.0),
                        "lon": data.get("lon", 0.0),
                        "timezone": data.get("timezone", ""),
                        "isp": data.get("isp", "Unknown"),
                        "org": data.get("org", "Unknown"),
                        "as": data.get("as", ""),
                        "asName": data.get("asname", ""),
                        "reverse": data.get("reverse", ""),
                        "isMobile": bool(data.get("mobile")),
                        "isProxy": bool(data.get("proxy")),
                        "isHosting": bool(data.get("hosting")),
                        "riskIndicators": _assess_ip_risk(data),
                        "error": None,
                    }
                else:
                    return _create_fallback_result(ip, data.get("message", "Lookup failed"))
            else:
                return _create_fallback_result(ip, f"HTTP status {response.status_code}")
    except Exception as e:
        return _create_fallback_result(ip, str(e))

async def batch_lookup_ips(ips: list) -> list:
    results = []
    for i, ip in enumerate(ips):
        if i > 0:
            await asyncio.sleep(0.3)
        res = await lookup_ip(ip)
        results.append(res)
    return results

def _assess_ip_risk(data: dict) -> list:
    indicators = []

    if data.get("proxy"):
        indicators.append({
            "type": "PROXY_VPN",
            "severity": "high",
            "description": "IP is associated with a proxy, VPN, or anonymization gateway"
        })

    if data.get("hosting"):
        indicators.append({
            "type": "HOSTING_PROVIDER",
            "severity": "medium",
            "description": f"IP belongs to a hosting/cloud provider ({data.get('org') or data.get('isp')})"
        })

    cloud_provider = _detect_cloud_provider(data.get("org", ""), data.get("isp", ""), data.get("as", ""))
    if cloud_provider:
        indicators.append({
            "type": "CLOUD_INFRASTRUCTURE",
            "severity": "medium",
            "description": f"IP hosted on {cloud_provider} infrastructure"
        })

    if _is_tor_indicator(data.get("org", ""), data.get("isp", ""), data.get("reverse", "")):
        indicators.append({
            "type": "TOR_EXIT",
            "severity": "critical",
            "description": "IP matches known TOR exit node or relay signature"
        })

    high_risk_countries = ['RU', 'CN', 'KP', 'IR', 'NG', 'RO', 'UA', 'BY']
    if data.get("countryCode") in high_risk_countries:
        indicators.append({
            "type": "HIGH_RISK_COUNTRY",
            "severity": "medium",
            "description": f"IP originates from high-risk jurisdiction: {data.get('country')}"
        })

    return indicators

def _detect_cloud_provider(org: str, isp: str, as_field: str) -> str:
    combined = f"{org} {isp} {as_field}".lower()
    if 'amazon' in combined or 'aws' in combined or 'ec2' in combined:
        return 'Amazon Web Services (AWS)'
    if 'microsoft' in combined or 'azure' in combined:
        return 'Microsoft Azure'
    if 'google' in combined or 'gcp' in combined:
        return 'Google Cloud Platform'
    if 'digitalocean' in combined:
        return 'DigitalOcean'
    if 'linode' in combined or 'akamai' in combined:
        return 'Linode/Akamai'
    if 'ovh' in combined:
        return 'OVH'
    if 'hetzner' in combined:
        return 'Hetzner'
    if 'cloudflare' in combined:
        return 'Cloudflare'
    return ""

def _is_tor_indicator(org: str, isp: str, reverse: str) -> bool:
    combined = f"{org} {isp} {reverse}".lower()
    return 'tor' in combined or 'exit' in combined or 'relay' in combined

def _create_fallback_result(ip: str, err: str) -> dict:
    return {
        "ip": ip,
        "country": "Unknown",
        "countryCode": "XX",
        "region": "Unknown",
        "city": "Unknown",
        "lat": 0.0,
        "lon": 0.0,
        "timezone": "",
        "isp": "Unknown",
        "org": "Unknown",
        "as": "",
        "asName": "",
        "reverse": "",
        "isMobile": False,
        "isProxy": False,
        "isHosting": False,
        "riskIndicators": [],
        "error": err,
    }

def calculate_ip_risk_score(ip_result: dict) -> int:
    if not ip_result or ip_result.get("error"):
        return 30
    score = 0
    for ind in ip_result.get("riskIndicators", []):
        sev = ind.get("severity")
        if sev == "critical":
            score += 40
        elif sev == "high":
            score += 30
        elif sev == "medium":
            score += 15
        elif sev == "low":
            score += 5
    return min(100, score)
